qa/workunits/windows/test_rbd_wnbd.py

```diff
@@ -285,6 +285,13 @@ def _wait_for_disk(self,
     def path(self):
         return f"\\\\.\\PhysicalDrive{self.disk_number}"
 
+    @Tracer.trace
+    def set_writable(self):
+        execute(
+            "powershell.exe", "-command",
+            "Set-Disk", "-Number", str(self.disk_number),
+            "-IsReadOnly", "$false")
```

Comment:
Do we know why the disk is coming up read-only?

I'm not convinced that just fixing up the test is the right approach.  Would the user need to muck with setting `IsReadOnly` to false as well, whether via shell or GUI?

Reply:
There are a few things that might cause this:

* the SanPolicy setting: https://learn.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-partitionmanager-sanpolicy
* windows updates
* the fact that we're no longer reporting the bus as "virtual" but as "SAS", which might have different defaults. We had to do this because of Microsoft Failover Cluster, which wouldn't accept RBD disks otherwise

FWIW, Windows Server 2022 seems to work as expected. I could try to reproduce it with the CI image. TBH, I wasn't too concerned about it because I remember seeing similar behavior with iSCSI.

We should probably disable this test to get the Windows job back on track while I do a little more digging.

Reply:
> There are a few things that might cause this:
> 
>     * the SanPolicy setting: https://learn.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-partitionmanager-sanpolicy

FWIW I don't see anything concerning read-only state on that page.  It talks about either mounting or not mounting (a subset of) disks.

> 
>     * windows updates
> 
>     * the fact that we're no longer reporting the bus as "virtual" but as "SAS", which might have different defaults. We had to do this because of Microsoft Failover Cluster, which wouldn't accept RBD disks otherwise

Mounting a disk can involve writing to it even if the mount is read-only, e.g. to update "last mounted on" timestamp or similar.  Defaulting physically writable storage devices to read-only seems weird...

> 
> FWIW, Windows Server 2022 seems to work as expected. I could try to reproduce it with the CI image. TBH, I wasn't too concerned about it because I remember seeing similar behavior with iSCSI.

Yeah, I'd be curious to learn what is causing this.  Depending on the findings, one option might be to flip to read-write in the driver itself for all non-snapshot RBD mappings.

Reply:
I ran some tests and I can now confirm that this was caused by the bus type and the SAN policy setting.

Here's a better explanation of this setting: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/san: 

```
```

Reply:
> we should probably stick to the host configuration and avoid having rbd automatically enable the disk. We do have to properly document this though.

Definitely, including a PowerShell command to flip the disk to read-write.

One thing is still not clear to me: you said earlier that Windows Server 2022 seemed to work as expected for you.  Yet the linked page applies as far back as Windows Server 2012 and the default appears to have been `offlineShared` since before then (https://learn.microsoft.com/en-us/windows/win32/api/vds/ne-vds-vds_san_policy).  Could it be that you had a custom policy configured?

Reply:
Indeed, my WS 2022 host was using the "onlineAll" policy instead of "offlineShared". However, there's more to it.

Windows remembers the configured disk status, so if disk 1 is enabled and writable, it will remain so after a host reboot or disk reconnect. The fact that Windows remembers the disk status based on the disk number is troublesome though considering that the disk numbers can change after reconnecting the disks.

Here's an example in which the wrong disk becomes online after a service restart: https://pastebin.com/raw/Fn8tWZwP
Worth mentioning that this also affects iSCSI: https://pastebin.com/raw/2ddNaNhk

Partition letters are preserved after reconnecting the disks, however we're still hitting the same issue with the disk status: https://pastebin.com/raw/MacbQZJz

IMO this is another example of poor legacy Windows design and there's not much we can do about it other than document it. We had a similar issue with Hyper-V disk attachments: https://docs.ceph.com/en/quincy/rbd/rbd-windows/#hyper-v-disk-addressing.

I'll update the PR description and document the Windows san policy.

Reply:
OK, so given these findings, RBD disks are largely expected to be come up offline and read-only.  This change only unsets `IsReadOnly`, where is `IsOffline` expected to be unset for the test?

Reply:
We don't need to change the "IsOffline" flag when accessing the disk directly. Setting "IsOffline" to false will basically mount the partitions, which isn't the case here. I guess it's worth mentioning this in the documentation though.

Reply:
+1 for mentioning this in the doc, but it still feels like it should be done here as well just because `set_writable` call is added to the generic test harness, not to the specific test.  What if a future test wants to mount the partitions?

Reply:
Makes sense, I'll add that as well. 

> What if a future test wants to mount the partitions? Or would that conflict with raw disk access?
It depends on the test case, as long as we aren't performing raw writes to a disk that has partitions we should be fine.

I guess it can even be opt-out. This operation can take ~20 seconds under heavy load (~50 disks being attached at the same time) and can be skipped depending on the san policy.